snapcraft/internal/repo/_base.py

```diff
@@ -0,0 +1,196 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2017 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import contextlib
+import fileinput
+import itertools
+import logging
+import os
+import re
+import shutil
+import stat
+
+from snapcraft import file_utils
+from . import get_pkg_libs
+
+
+_BIN_PATHS = (
+    'bin',
+    'sbin',
+    'usr/bin',
+    'usr/sbin',
+)
+
+
+logger = logging.getLogger(__name__)
+
+
+class BaseRepo:
+    """Base implementation for a platform specific repo handler.
+
+    Generally any new repo handling system would inherit from this and
+    implement `get` and `unpack`. At the end of the `unpack` method
+    `normalize` needs to be called to adapt the artifacts downloaded
+    to be generic enough for building a snap."""
+    def __init__(self, rootdir, *args, **kwargs):
+        """Initialize a repository handler.
+
+        :param rootdir string: the root directory to work on. This may well be
+                               to create repo handling specific caches or
+                               download artifacts.
+        """
+        self.rootdir = rootdir
+
+    def get(self, package_names):
+        """Get the packages.
+
+        This method needs to be implemented by the inheriting class. It
+        is called all along the core of snapcraft to obtain the
+        package_names and store them in some repo defined location to
+        later unpack.
+
+        :param str package_names: list of packages to obtain.
```

Comment:
next time you ask me to do this, fetch out the syntax or get us going with markdown for docs :-)